Added support for large packet handling

kiseitai3 opened the pull request:
In FSK/OOK mode, packets sent or received by some devices can be larger than the size of FIFO. Per Section 4.2.13.5 of the datasheet of the SX127x family of radio modules, large packets are handled with an "on the fly" FIFO clearing/filling scheme. I implemented such schemes within both transmit and receive methods. The drivers should be able to handle transmission of large packets properly. This will allow the library to be used to communicate with devices that expect large packets and might not defragment large payloads broken into multiple packets.

kiseitai3 committed:
Added support for large packet handling

In FSK/OOK mode, packets sent or received by some devices can be larger than the size of FIFO. Per Section 4.2.13.5 of the SX127x family of radio modules, large packets are handled with an "on the fly" FIFO clearing/filling scheme. I implemented such schemes within both transmit and receive methods. The drivers should be able to handle transmission of large packets properly. This will allow the library to be used to communicate with devices that expect large packets and might not defragment large payloads broken into multiple packets.

kiseitai3 committed:
Hotfix: forgot to get address of array location (arr[i]) instead os contents

kiseitai3 committed:
Oops! Extra parenthesis

kiseitai3 commented:
I have a question. If I have understood the drivers correctly, does transmit and receive rely on the variable packet format? Is there a chance that I could implement a transmit/receive version that works under the assumption of the Unlimited Packet Format? The way I would implement it in the following days would be by adding an argument packetType (set to Variable Packet Format via a macro by default) to transmit and receive, so they can call the appropriate private methods to which I would move the current transmit/receive code. As a bonus, I am willing to add support for the Fixed packet format (Section 4.2.13.2). I think I may need this bit (unlimited packet format) for my project and I figured this would be the cleanest way to do it. My plan B was to do some macro magic to build an inheritance scheme at compile time that would give my new set of methods access to the underlying spi write methods and infrastructure. The inheritance scheme was going to be implemented in my project rather than LoRaLib, but I find that approach a bit hacky. At the end of the day, I understand if you reject this added proposal. If you allow me to implement and publish this last bit of changes, don't close this PR so I can just package everything prior to release (unless your projects are on a fixed release timeline). By the way, thank you for building this library and for generously allowing me to submit my previous patches! :)

jgromes commented on src/modules/SX127x.cpp:
```diff
@@ -327,7 +363,31 @@ int16_t SX127x::receive(uint8_t* data, size_t len) {
       data = new uint8_t[length + 1];
     }
-    _mod->SPIreadRegisterBurst(SX127X_REG_FIFO, length, data);
+    else{
```
Why is the actual reception code in else branch? It won't be able to receive anything if user calls it with String class (len parameter equal to zero).

jgromes commented on src/modules/SX127x.cpp:
```diff
@@ -327,7 +363,31 @@ int16_t SX127x::receive(uint8_t* data, size_t len) {
+            break;
+          }
+          //Read a single byte at a time if the radio module has not finished receiving!
+          _mod->SPIreadRegisterBurst(SX127X_REG_FIFO, 1, &data[rcvByteCount]);
```
No need to use burst SPI access when reading single byte

jgromes commented on src/modules/SX127x.h:
```diff
@@ -226,6 +226,7 @@
 #define SX127X_MODULATION_FSK                         0b00000000  //  6     5     FSK modulation scheme
 #define SX127X_MODULATION_OOK                         0b00100000  //  6     5     OOK modulation scheme
 #define SX127X_RX                                     0b00000101  //  2     0     receiver mode
+#define SX127X_FSK_OOK_FIFOLIMIT   0x40  //64 byte limit on payload slice that FIFO accepts
```
This macro shouldn't be in section related to the REG_OP_MODE register. Move to REG_PAYLOAD_LENGTH_FSK section (or discard the macro entirely and read the value from REG_PAYLOAD_LENGTH_FSK register).

kiseitai3 replied:
The macro doesn't feel like it belongs anywhere, but it's a magic number. I thought the register Reg_Payload_Length is just a place holder for the desired packet length or receive packet length. Is this a completely different register? If so, I'll just read from it. I don't know which constants are stored in hardware and which we have to define on our own using the datasheet.

jgromes replied:
All the macros are actual definitions for register addresses or variables - there are no "magic numbers". The way the register definitions are strucutured is described [here](https://github.com/jgromes/RadioLib/blob/master/extras/ModuleTemplate.h)

kiseitai3 replied:
There may not be any defined magic numbers in the header, but the implementation is littered with them. For example, 32000 in a few locations for the crystal oscillator frequency. That's not even true either. There's a macro at the top of the header for the crystal oscillator frequency that defines a float => 32.0. The fact that we know the FIFO has a size x in hardware makes it a magic number by definition. The question is whether the hardware has those magic numbers burnt in it (firmware or silicone), or if we have to define it somewhere like I did so we can optimize transfer. The datasheet says PayloadLength will return max size under variable length packet mode, so you may be right about the possibility of simply reading the register and storing it ahead of time before tx_mode time but only under variable packet mode. What I meant about Reg_Payload_Length being a place holder is that whether the hardware register PayloadLength holds the size of received packets and the size (by setting it in transmit) of packet to be transmitted. 

I agree I probably should read the register. The datasheet says it is not used in transmit mode, so I am not sure if the register would return the desired value at all. (page 100, RegPayloadLength at the bottom of the table, datasheet)

jgromes replied:
I agree that there's a lot of unexplained constants in the implementation, my original remark was about the structure of macros in header file.

jgromes commented on src/modules/SX127x.cpp:
```diff
@@ -327,7 +363,31 @@ int16_t SX127x::receive(uint8_t* data, size_t len) {
+            break;
+          }
+          //Read a single byte at a time if the radio module has not finished receiving!
+          _mod->SPIreadRegisterBurst(SX127X_REG_FIFO, 1, &data[rcvByteCount]);
```
No need to use burst SPI access when reading single byte

kiseitai3 replied:
This was me being consistent. Changed that in my local version.

jgromes commented on src/modules/SX127x.cpp:
```diff
@@ -327,7 +363,31 @@ int16_t SX127x::receive(uint8_t* data, size_t len) {
     }
-    _mod->SPIreadRegisterBurst(SX127X_REG_FIFO, length, data);
+    else{
+      while(rcvByteCount < length){
```
Why is the actual reception code in else branch? It won't be able to receive anything if user calls it with String class (len parameter equal to zero).

kiseitai3 replied:
That one was my brain interpreting the inital part as "here's an empty buffer since you did not provide a buffer of appropriate size." Totally my mistake.

jgromes commented:
Were you able to test it with hardware? I can't seem to get it to work when sending packets longer than 64 bytes - it does send out something FSK-modulated, but then it gets stuck in an infinite loop in the `transmit()` method.

Also, the `receive()` method doesn't work for String calls, even for packets shorter than 64 bytes.

As per your questions: yes, the current FSK receive/transmit is using variable length format. What would be the use case for unlimited/fixed packets? I'm generally against implementing something "just 'cause", especially if it increases overall size of the library and complexity on the user side.

kiseitai3 commented:
I do not have the test equipment, so unfortunately I can't test the changes right now. 

How many bytes are you receiving during transmit? I think I know why the loops are not breaking. I thought SX127X_FLAG_FIFO_EMPTY was the actual register. I think I was supposed to call IRQ_FLAGS and mask it with SX127X_FLAG_FIFO_EMPTY. Sorry for my silly mistake.

As per the other request, the use case is having the ability to talk to a proprietary device for which you may not have perfect imformation about the packet, or the company uses a proprietary packet. Both fixed and unlimited length packet formats allow you to get the bare bones packet data that I can then parse in my project. I understand the size issue since my project also has that constraint (both firmware and ram). However, I do understand that I have the liberty of using macro magic to skip compiling certain modules if they are not going to be used, which is not the case for a library. If you say no, no worries. I just get to go the hacky route. :P 

P.S. Yeah, I was supposed to call IRQ_FLAGS_2. I'm sorry about this.

kiseitai3 commented:
I was discussing with a colleague what packet modes I would have to support in my own project and he told me both unlimited and fixed. The target hardware sends arbitrary sized messages (Unlimited). We have to talk back to the hardware without postamble bit glitches and the only format that stops transmitting in hardware after the last byte is the fixed format. The final packets are then correctly interpreted and accepted by the target hardware.

kiseitai3 committed:
Fixed transmit and receive

transmit: Fixed improper reading of flags.
receive: Fixed improper reading of register flags, wrong else clause, and improper use of Burst method for one byte reads.

Extra: Standardized setBit math, although it should not change results.

kiseitai3 committed:
Merge branch 'master' into master

kiseitai3 committed:
Fixed missing bracket during synch

kiseitai3 commented:
There's one thing I don't understand from the datasheet, what do they mean by BitRate(15,0) in the Bitrate calculation.

jgromes commented:
It means bit rate as 16-bit number.

kiseitai3 commented:
Hmm, so if we really wanted to follow the datasheet to the letter, it asks us to convert the float br into int16_t, right? I guess that bit of code is ok as is. :)

jgromes commented:
That's what the calculation does - it turns the float bit rate in kbps to the 16-bit raw bitrate value that can be written into register. BitRate(15,0) is the register value.

kiseitai3 commented:
You are right.

jgromes commented:
Also, it still doesn't seem to be working for packets larger than 64 bytes. I *think* that the transmission is fine (at least from the SDR output), but nothing is ever received.

From experience, I know how hard it is to write code without having access to the hardware. With that in mind, I think it might be best to put this aside for now, release v8.0.0 with OOK and wait until you get the hardware so that you can test directly.

As for the fixed/unlimited packets, I'm not against it if there's a use for it if; we can have separate methods (e.g. `transmitFixed()` and `transmitUnlimited()`) or a new method `setPacketType().` There's already one optional parameter for `transmit()`, and that's the address. Addresses are independent of packet type, so there could be a conflict between different optional parameters - most users will want to transmit variable packets without address filtering (at least imo).

kiseitai3 commented:
I agree, then. Unfortunately, I don't know when I will be able to get the hardware. I'm unemployed at the moment. Thank you for giving it a shot!

jgromes commented:
Closed due to inactivity, feel free to reopen later.
